src/protocol/boolean/saturating_sum.rs

```diff
@@ -0,0 +1,170 @@
+    where
+        C: Context,
+        S: LinearSecretSharing<Gf2> + BasicProtocols<C, Gf2>,
+    {
+        assert!(self.sum.len() >= value.len());
+
+        let mut output_sum = Vec::with_capacity(self.sum.len());
+        let mut carry_in = S::ZERO;
+        let zero = S::ZERO;
+        for i in 0..self.sum.len() {
+            let c = ctx.narrow(&BitOpStep::from(i));
+            let x = if i < value.len() { &value[i] } else { &zero };
+            let (sum_bit, carry_out) =
+                one_bit_adder(c, record_id, x, &self.sum[i], &carry_in).await?;
+
+            output_sum.push(sum_bit);
+            carry_in = carry_out;
+        }
+        let is_saturated = -carry_in
+            .clone()
+            .multiply(
+                &self.is_saturated,
+                ctx.narrow(&BitOpStep::from(self.sum.len())),
+                record_id,
+            )
+            .await?
+            + &carry_in
+            + &self.is_saturated;
+
+        Ok(SaturatingSum::new(
+            BitDecomposed::new(output_sum),
+            is_saturated,
+        ))
+    }
+}
+
+///
+/// Returns (`sum_bit`, `carry_out`)
+///
+async fn one_bit_adder<C, SB>(
+    ctx: C,
+    record_id: RecordId,
+    x: &SB,
+    y: &SB,
+    carry_in: &SB,
+) -> Result<(SB, SB), Error>
+where
+    C: Context,
+    SB: LinearSecretSharing<Gf2> + BasicProtocols<C, Gf2>,
+{
+    // compute sum bit as x XOR y XOR carry_in
+    let sum_bit = x.clone() + y + carry_in;
+
+    let x_xor_carry_in = x.clone() + carry_in;
+    let y_xor_carry_in = y.clone() + carry_in;
+    let carry_out = x_xor_carry_in
+        .multiply(&y_xor_carry_in, ctx, record_id)
+        .await?
+        + carry_in;
+
+    Ok((sum_bit, carry_out))
+}
+
+#[cfg(all(test, unit_test))]
+mod tests {
+    use super::SaturatingSum;
+    use crate::{
+        ff::{Field, Gf2},
+        protocol::{context::Context, RecordId},
+        secret_sharing::{
+            replicated::semi_honest::AdditiveShare as Replicated, BitDecomposed, SharedValue,
+        },
+        test_fixture::{get_bits, Reconstruct, Runner, TestWorld},
+    };
+
+    impl Reconstruct<u128> for [SaturatingSum<Replicated<Gf2>>; 3] {
+        fn reconstruct(&self) -> u128 {
+            let [s0, s1, s2] = self;
+
+            let sum_bits: BitDecomposed<Gf2> = BitDecomposed::new(
+                s0.sum
+                    .iter()
+                    .zip(s1.sum.iter())
+                    .zip(s2.sum.iter())
+                    .map(|((a, b), c)| [a, b, c].reconstruct()),
+            );
+
+            let is_saturated = [&s0.is_saturated, &s1.is_saturated, &s2.is_saturated].reconstruct();
+
+            if is_saturated == Gf2::ZERO {
+                sum_bits
+                    .iter()
+                    .map(Field::as_u128)
+                    .enumerate()
+                    .fold(0_u128, |acc, (i, x)| acc + (x << i))
+            } else {
+                2_u128.pow(s0.sum.len() as u32)
```

Comment:
You should assert that all of the bits are also 1 when saturation occurs.

Reply:
But they are not. That would be more costly. The less significant bits continue to change as you keep accumulating more numbers - but the proper interpretation is to ignore them when the "is_saturated" bit is a 1